components/script/task_manager.rs

```diff
@@ -9,35 +9,22 @@ use std::sync::Arc;
 use base::id::PipelineId;
 
 use crate::dom::bindings::cell::DomRefCell;
-use crate::messaging::MainThreadScriptChan;
+use crate::script_runtime::ScriptChan;
 use crate::task::TaskCanceller;
-use crate::task_source::dom_manipulation::DOMManipulationTaskSource;
-use crate::task_source::file_reading::FileReadingTaskSource;
-use crate::task_source::gamepad::GamepadTaskSource;
-use crate::task_source::history_traversal::HistoryTraversalTaskSource;
-use crate::task_source::media_element::MediaElementTaskSource;
-use crate::task_source::networking::NetworkingTaskSource;
-use crate::task_source::performance_timeline::PerformanceTimelineTaskSource;
-use crate::task_source::port_message::PortMessageQueue;
-use crate::task_source::remote_event::RemoteEventTaskSource;
-use crate::task_source::rendering::RenderingTaskSource;
-use crate::task_source::timer::TimerTaskSource;
-use crate::task_source::user_interaction::UserInteractionTaskSource;
-use crate::task_source::websocket::WebsocketTaskSource;
-use crate::task_source::TaskSourceName;
+use crate::task_source::{TaskSource, TaskSourceName};
 
 macro_rules! task_source_functions {
-    ($self:ident, $task_source:ident, $task_source_type:ident, $task_source_name:ident) => {
-        pub(crate) fn $task_source(&$self) -> $task_source_type {
+    ($self:ident, $task_source:ident) => {
+        pub(crate) fn $task_source(&$self) -> TaskSource {
```

Comment:
Not a blocker for this PR, just something I noticed  - in a lot of cases, the task source returned by this method is only ever used for calling `.queue_*()` on it immediately. This method both clones the channel and allocates a new Box and based on adding `println` in the clone method, it seems like this is happening quite a lot. 

Is there a plan to eventually use a single instance of task source, or would that introduce issues related to message ordering?

Reply:
Yeah, this is pretty inefficient. A lot of times the `TaskSource` is moved into handlers which requires a clone anyway, but it would be nice to only clone when necessary. I hope we can make this change once the canceller is part of the `TaskSource` itself...which should make the API a lot more ergonomic in general.